### ISSUE TITLE
theano.tensor.switch

### ISSUE THREAD

**Issue by ghost** [Theano/Theano#3220]
Hi,

I have the following problem:
I expect the following line of code to grab values from a 3d tensor when a condition is met and place it in its corresponding location in "output" otherwise put a zero in that location in "output".

output = T.switch( T.ge(y,input_mat.shape[1]) or T.lt(y,0.0) or T.ge(x,input_mat.shape[2]) or
                             T.lt(x,0.0), 0.0, input_mat[idx, y, x]  )

where "y" is a matrix of size (batch, dim) containing indices, "x" is also a matrix of size (batch, dim) containing indices, input_mat is of size (batch, height, width), height > dim, width > dim, output is of size (batch, dim) after computation, and idx is of size (batch, dim) where idx is just a vector of indices for batch index repeated "dim" times along axis=1. The problem I am having is that my condition, somehow, evaluates to false for all elements and therefore always grabs the values from "input_mat". What I am trying to do here is prevent my indices from being out of bounds while if that was to happen to just put a zero in the corresponding location of "output". Any help would really be appreciated, If I am not seeing something obvious in my condition please point it out, and if there is anything I am misunderstanding from the T.switch function also please let me know.

Thanks in advance!

**Comment by escopeta** [comment]
UPDATE: I solved my problem with more lines of code, but I would still like to see how this can be done with switch which would result in less lines. Thanks

**Comment by nouiz** [comment]
Can you give the code that do what you want? Sometimes it is easier to understand code, there is no ambiguity in what it does.

**Comment by escopeta** [comment]
bound_y1    = T.ge(y, input.shape[1]).nonzero()
bound_y2    = T.lt(y, 0.0).nonzero()

bound_x1   = T.ge(x, input.shape[2]).nonzero()
bound_x2   = T.lt(x, 0.0).nonzero()

# this step is for just picking the pixel at 0.0 which will then be replaced
y                 = T.set_subtensor(y[bound_y1], 0.0)
y                 = T.set_subtensor(y[bound_y2], 0.0)
x                 = T.set_subtensor(x[bound_x1], 0.0)
x                 = T.set_subtensor(x[bound_x2], 0.0)

output         = U[idx, y, x]
output         = T.set_subtensor(U_topleft[bound_y1], 0.0)
output         = T.set_subtensor(U_topleft[bound_y2], 0.0)
output         = T.set_subtensor(U_topleft[bound_x1], 0.0)
output         = T.set_subtensor(U_topleft[bound_x2], 0.0)


I'm sorry if it's too hacky :)

**Comment by escopeta** [comment]
I think I have made it work, what the problem I was having is that I would get the index out of bounds error even after I identified the conditions. While using switch, it seems it will evaluate the ifthen and else arguments before making the decision of which to pick based on the condition. Because of this, the following hack works for me:


output = T.switch( T.ge(y,input_mat.shape[1]) or T.lt(y,0.0) or T.ge(x,input_mat.shape[2]) or
T.lt(x,0.0), 0.0, input_mat[idx, T.clip(y,0.0,input_mat.shape[1]), T.clip(x,0.0,input_mat.shape[2])] )

I just made sure to clip the bounds for the values to return after the conditions have been observed and I still get the value I wanted which in the clipping case would be the 0.0 in the code I put here.
